README.rst
##############
pyramid_walrus
##############

a `walrus`_ integration for pyramid


To install::

        $ pip install pyramid_walrus

In your configuration::

        [app:main]

        ...

        pyramid.includes =
            pyramid_walrus


In your views::

        def my_view(request):
            db = request.walrus_db
            db.incr('my_counter')


Backend-specific configurations
===============================


`walrus`_ can be configured to use several `redis`_-like backends.


`Redis`_ (default)
------------------

::

        $ pip install redis

::

        pyramid_walrus.backend = redis
        pyramid_walrus.backend.redis.url = redis://@localhost:6379/0
        pyramid_walrus.backend.redis.max_connections = 10


`Rlite`_
--------

::

        $ pip install hirlite

::

        pyramid_walrus.backend = rlite
        pyramid_walrus.backend.rlite.filename = :memory:
        pyramid_walrus.backend.rlite.encoding = utf-8

`Vedis`_
--------

::

        $ pip install vedis

::

        pyramid_walrus.backend = vedis
        pyramid_walrus.backend.vedis.filename = :memory:

`LedisDB`_
----------

::

        $ pip install ledis

::

        pyramid_walrus.backend = ledis
        pyramid_walrus.backend.ledis.url = ledis://@localhost:6380/0
        pyramid_walrus.backend.ledis.max_connections = 10


Custom
------

::

        pyramid_walrus.backend = myapp.get_walrus_backend_from_settings


and define your custom factory function::

        # myapp.py

        def get_walrus_backend_from_settings(settings):
            def get_database_for_request(request):
                ...
                return db
            return get_database_for_request

.. _walrus: https://pypi.python.org/pypi/walrus
.. _redis: http://redis.io
.. _rlite: https://github.com/seppo0010/rlite
.. _Vedis: http://vedis.symisc.net/
.. _LedisDB: http://ledisdb.com/



CHANGES.rst
Changes
=======

0.2.0 (unreleased)
------------------

- Nothing changed yet.


0.1.2 (2015-06-06)
------------------

- Fix COPYING.


0.1.1 (2015-06-06)
------------------

- Change license to GNU Lesser General Public License v3 or later (LGPLv3+)
- Fix packaging.


0.1.0 (2015-06-05)
------------------

- Initial release.


pyramid_walrus/tests/__init__.py
# -*- coding: utf-8 -*-
#
#   pyramid_walrus : a walrus integration for pyramid
#   Copyright (C) 2015 mete0r <mete0r@sarangbang.or.kr>
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
from __future__ import unicode_literals


pyramid_walrus/tests/test_backends.py
# -*- coding: utf-8 -*-
#
#   pyramid_walrus : a walrus integration for pyramid
#   Copyright (C) 2015 mete0r <mete0r@sarangbang.or.kr>
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
from __future__ import unicode_literals
from unittest import TestCase
import os.path

from ..backends import get_backend_factory
from ..backends import RedisFactory
from ..backends import RLiteFactory
from ..backends import VedisFactory
from ..backends import LedisFactory


class get_backend_factory_Test(TestCase):

    def get_db(self, settings):
        factory = get_backend_factory(settings)
        request = None
        return factory(request)

    def test_default(self):
        from walrus import Database
        settings = {}
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, Database))

    def test_redis(self):
        from walrus import Database
        settings = {
            'pyramid_walrus.backend': 'redis'
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, Database))
        self.assertEquals(10, db.connection_pool.max_connections)

    def test_rlite(self):
        from walrus.tusks.rlite import WalrusLite
        settings = {
            'pyramid_walrus.backend': 'rlite'
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusLite))

    # XXX: disabled
    def xxx_test_vedis(self):
        from walrus.tusks.vedisdb import WalrusVedis
        settings = {
            'pyramid_walrus.backend': 'vedis'
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusVedis))

    # XXX: disabled for broken walrus-0.3.1 + ledis-0.0.1 in PyPI
    def xxx_test_ledis(self):
        from walrus.tusks.ledisdb import WalrusLedis
        settings = {
            'pyramid_walrus.backend': 'ledis'
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusLedis))
        self.assertEquals(10, db.connection_pool.max_connections)

    def test_custom(self):
        settings = {
            'pyramid_walrus.backend':
            'pyramid_walrus.tests.test_backends.get_custom_backend_factory'
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, CustomDatabase))


def get_custom_backend_factory(settings):
    def get_database_for_request(request):
        return CustomDatabase()
    return get_database_for_request


class CustomDatabase(object):
    pass


class RedisFactoryTest(TestCase):

    def get_db(self, settings):
        get_database_for_request = RedisFactory(settings)
        db = get_database_for_request(None)
        return db

    def test_db1(self):
        from redis.connection import Connection
        from redis.connection import BlockingConnectionPool
        from walrus import Database
        settings = {
            'pyramid_walrus.backend.redis.url': 'redis://@127.0.0.1:6379/1',
            'pyramid_walrus.backend.redis.max_connections': '5',
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, Database))

        connection_pool = db.connection_pool
        self.assertTrue(isinstance(connection_pool, BlockingConnectionPool))
        self.assertEquals(5, connection_pool.max_connections)

        kwargs = connection_pool.connection_kwargs
        self.assertEquals('127.0.0.1', kwargs['host'])
        self.assertEquals(6379, kwargs['port'])
        self.assertEquals(1, kwargs['db'])
        self.assertEquals(None, kwargs['password'])
        self.assertEquals(Connection, connection_pool.connection_class)

    def test_unixsocket(self):
        from redis.connection import UnixDomainSocketConnection
        from redis.connection import BlockingConnectionPool
        from walrus import Database
        settings = {
            'pyramid_walrus.backend.redis.url': 'unix://@/redis.sock?db=1'
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, Database))

        connection_pool = db.connection_pool
        self.assertTrue(isinstance(connection_pool, BlockingConnectionPool))

        kwargs = connection_pool.connection_kwargs
        self.assertEquals(1, kwargs['db'])
        self.assertEquals(None, kwargs['password'])
        self.assertEquals(UnixDomainSocketConnection,
                          connection_pool.connection_class)


class RLiteFactoryTest(TestCase):

    def setUp(self):
        self.diskfilename = self.id()
        if os.path.exists(self.diskfilename):
            os.unlink(self.diskfilename)

    def get_db(self, settings):
        get_database_for_request = RLiteFactory(settings)
        db = get_database_for_request(None)
        return db

    def test_memory(self):
        from walrus.tusks.rlite import WalrusLite
        settings = {
            'pyramid_walrus.backend.rlite.filename': ':memory:',
            'pyramid_walrus.backend.rlite.encoding': 'utf-8',
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusLite))
        self.assertEquals(':memory:', db._filename)
        self.assertEquals('utf-8', db._encoding)

    def test_diskfile(self):
        from walrus.tusks.rlite import WalrusLite
        settings = {
            'pyramid_walrus.backend.rlite.filename': self.diskfilename,
            'pyramid_walrus.backend.rlite.encoding': 'utf-8',
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusLite))
        self.assertEquals(self.diskfilename, db._filename)
        self.assertEquals('utf-8', db._encoding)


# XXX VedisFactoryTest(TestCase): disabled
class VedisFactoryTest(object):

    def setUp(self):
        self.diskfilename = self.id()
        if os.path.exists(self.diskfilename):
            os.unlink(self.diskfilename)

    def get_db(self, settings):
        get_database_for_request = VedisFactory(settings)
        db = get_database_for_request(None)
        return db

    def test_memory(self):
        from walrus.tusks.vedisdb import WalrusVedis
        settings = {
            'pyramid_walrus.backend.vedis.filename': ':memory:',
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusVedis))
        self.assertEquals(':memory:', db._filename)

    def test_diskfile(self):
        from walrus.tusks.vedisdb import WalrusVedis
        settings = {
            'pyramid_walrus.backend.vedis.filename': self.diskfilename,
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusVedis))
        self.assertEquals(self.diskfilename, db._filename)


# XXX LedisFactoryTest(TestCase): broken walrus-0.3.1 + ledis-0.0.1 in PyPI
class LedisFactoryTest(object):

    def get_db(self, settings):
        get_database_for_request = LedisFactory(settings)
        db = get_database_for_request(None)
        return db

    def test_db1(self):
        from ledis.connection import Connection
        from ledis.connection import BlockingConnectionPool
        from walrus.tusks.ledisdb import WalrusLedis
        settings = {
            'pyramid_walrus.backend.ledis.url': 'ledis://@127.0.0.1:6380/1',
            'pyramid_walrus.backend.ledis.max_connections': '5',
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusLedis))

        connection_pool = db.connection_pool
        self.assertTrue(isinstance(connection_pool, BlockingConnectionPool))
        self.assertEquals(5, connection_pool.max_connections)

        kwargs = connection_pool.connection_kwargs
        self.assertEquals('127.0.0.1', kwargs['host'])
        self.assertEquals(6380, kwargs['port'])
        self.assertEquals(1, kwargs['db'])
        self.assertEquals(None, kwargs['password'])
        self.assertEquals(Connection, connection_pool.connection_class)

    def test_unixsocket(self):
        from ledis.connection import UnixDomainSocketConnection
        from ledis.connection import BlockingConnectionPool
        from walrus.tusks.ledisdb import WalrusLedis
        settings = {
            'pyramid_walrus.backend.ledis.url': 'unix://@/redis.sock?db=1',
        }
        db = self.get_db(settings)
        self.assertTrue(isinstance(db, WalrusLedis))

        connection_pool = db.connection_pool
        self.assertTrue(isinstance(connection_pool, BlockingConnectionPool))

        kwargs = connection_pool.connection_kwargs
        self.assertEquals(1, kwargs['db'])
        self.assertEquals(None, kwargs['password'])
        self.assertEquals(UnixDomainSocketConnection,
                          connection_pool.connection_class)


pyramid_walrus/__init__.py
# -*- coding: utf-8 -*-
#
#   pyramid_walrus : a walrus integration for pyramid
#   Copyright (C) 2015 mete0r <mete0r@sarangbang.or.kr>
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
import logging


__version__ = '0.1.2'


logger = logging.getLogger(__name__)


def includeme(config):
    from .backends import get_backend_factory

    settings = config.registry.settings

    factory = get_backend_factory(settings)
    factory = wrap_factory(factory)
    config.add_request_method(factory, 'walrus_db', reify=True)


def wrap_factory(factory):
    def get_database_for_request(request):
        db = factory(request)
        assert db is not None
        return db
    return get_database_for_request


pyramid_walrus/backends.py
# -*- coding: utf-8 -*-
#
#   pyramid_walrus : a walrus integration for pyramid
#   Copyright (C) 2015 mete0r <mete0r@sarangbang.or.kr>
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
from __future__ import unicode_literals
import logging

from pyramid.path import DottedNameResolver

logger = logging.getLogger(__name__)


def get_backend_factory(settings):
    backend = settings.get('pyramid_walrus.backend', 'redis')
    if backend == 'redis':
        return RedisFactory(settings)
    elif backend == 'ledis':
        return LedisFactory(settings)
    elif backend == 'rlite':
        return RLiteFactory(settings)
    elif backend == 'vedis':
        return VedisFactory(settings)

    # custom factory
    resolver = DottedNameResolver()
    custom_factory = resolver.resolve(backend)
    return custom_factory(settings)


def RedisFactory(settings):
    from redis import BlockingConnectionPool
    from walrus import Walrus

    default = get_settings('pyramid_walrus.backend.redis.', settings, {
        'url': 'redis://@localhost:6379/0',
        'max_connections': '10',
    })

    connection_url = default['url']
    max_connection = default['max_connections']
    max_connection = int(max_connection) if max_connection else None
    connection_pool = BlockingConnectionPool.from_url(
        url=connection_url,
        max_connections=max_connection)
    logger.debug('connection_pool: %r', connection_pool)

    def get_database_for_request(request):
        return Walrus(connection_pool=connection_pool)
    return get_database_for_request


def LedisFactory(settings):
    from ledis import BlockingConnectionPool
    from walrus.tusks.ledisdb import WalrusLedis

    default = get_settings('pyramid_walrus.backend.ledis.', settings, {
        'url': 'ledis://@localhost:6380/0',
        'max_connections': '10',
    })
    connection_url = default['url']
    max_connection = default['max_connections']
    max_connection = int(max_connection) if max_connection else None
    connection_pool = BlockingConnectionPool.from_url(
        url=connection_url,
        max_connections=max_connection)
    logger.debug('connection_pool: %r', connection_pool)

    def get_database_for_request(request):
        return WalrusLedis(connection_pool=connection_pool)
    return get_database_for_request


def RLiteFactory(settings):
    from walrus.tusks.rlite import WalrusLite

    default = get_settings('pyramid_walrus.backend.rlite.', settings, {
        'filename': ':memory:',
        'encoding': 'utf-8',
    })
    db = WalrusLite(**default)

    def get_database_for_request(request):
        return db
    return get_database_for_request


def VedisFactory(settings):
    from walrus.tusks.vedisdb import WalrusVedis

    default = get_settings('pyramid_walrus.backend.vedis.', settings, {
        'filename': ':memory:',
    })
    db = WalrusVedis(**default)

    def get_database_for_request(request):
        return db
    return get_database_for_request


def get_settings(prefix, settings, default):
    for k in default:
        if prefix + k in settings:
            default[k] = settings[prefix + k]

    return default
